src/main/java/com/streamr/client/StreamrClient.java

```diff
@@ -232,29 +234,30 @@ public void connect() throws ConnectionTimeoutException {
             throw new ConnectionTimeoutException(options.getWebsocketApiUrl());
         }
 
-        if (inbox != null && inboxSub == null) {
-            inboxSub = subscribe(inbox, new MessageHandler() {
+        if (keyExchangeStream != null && keyExchangeSub == null) {
+            keyExchangeSub = subscribe(keyExchangeStream, new MessageHandler() {
                 @Override
                 public void onMessage(Subscription sub, StreamMessage message) {
                     try {
                         if (message.getContentType().equals(StreamMessage.ContentType.GROUP_KEY_REQUEST)) {
-                            keyExchangeUtil.handleGroupKeyRequest(message);
+                            try {
+                                keyExchangeUtil.handleGroupKeyRequest(message);
+                            } catch (Exception e) {
+                                GroupKeyRequest groupKeyRequest = GroupKeyRequest.fromMap(message.getContent());
+                                StreamMessage errorMessage = msgCreationUtil.createGroupKeyErrorResponse(message.getPublisherId(), groupKeyRequest, e);
+                                publish(errorMessage); //sending the error to the sender of 'message'
+                            }
                         } else if (message.getContentType().equals(StreamMessage.ContentType.GROUP_KEY_RESPONSE_SIMPLE)) {
                             keyExchangeUtil.handleGroupKeyResponse(message);
                         } else if (message.getContentType().equals(StreamMessage.ContentType.GROUP_KEY_RESET_SIMPLE)) {
                             keyExchangeUtil.handleGroupKeyReset(message);
-                        } else if (message.getContentType().equals(StreamMessage.ContentType.ERROR_MSG)) {
+                        } else if (message.getContentType().equals(StreamMessage.ContentType.GROUP_KEY_RESPONSE_ERROR)) {
                             handleInboxStreamErrorMessage(message);
                         } else {
                             throw new MalformedMessageException("Cannot handle message with content type: " + message.getContentType());
                         }
                     } catch (Exception e) {
-                        log.warn(e.getMessage());
-                        // we don't notify the error to the originator if the message is unauthenticated.
-                        if (message.getSignature() != null) {
-                            StreamMessage errorMessage = msgCreationUtil.createErrorMessage(message.getPublisherId(), e);
-                            publish(errorMessage); //sending the error to the sender of 'message'
-                        }
+                        log.error(e.getMessage());
```

Comment:
Should it "notify the error to the originator"

Reply:
According to the updated spec, the error message is a failure response to the `GroupKeyRequest` (only). I'm not sure if it makes sense to send error responses to the responses. The old code was doing it though, and also saw that in the JS client. Maybe needs a bit of discussion what the spec should be.

Reply:
Update: we discussed this and agreed that let's send `GroupKeyErrorResponse` only as a response to `GroupKeyRequest`, but not as a response to any of the responses. Resolving.